Document the binary protocol (#642)
edgedb/edgedb: docs/internals/protocol/dataformats.rst

        // Dimension data.
        Dimension   dimensions[ndims];
        // Element data, the number of elements
        // in this array is the sum of dimension sizes:
        // sum((d.upper - d.lower + 1) for d in dimensions)
        Element     elements[];
    };

    struct Dimension {
        // Upper dimension bound, inclusive,
        // number of elements in the dimension
        // relative to the lower bound.
        int32       upper;
        // Lower dimension bound, always 1.
        int32       lower;
    };

    struct Element {
        // Encoded element data length in bytes.
        int32       length;
        // Element data.
        byte        data[length];
    };

.. _ref_protocol_fmt_tuple:

tuple<>
=======

The tuple values are represented as the following structure:

.. code-block:: c

    struct TupleValue {
        // Number of elements
        int32       nelems;
        // Element data.
        Element     elements[nelems];
    };

    struct Element {
        // Encoded element data length in bytes.
        int32       length;
        // Element data.
        byte        data[length];
    };

.. _ref_protocol_fmt_uuid:

std::uuid
=========

> The :eql:type:`std::uuid` values are represented as an array of 16 unsigned

Reviewer: ```suggestion
The :eql:type:`std::uuid` values are represented as a sequence of 16 unsigned
```

    };

The decimal values are represented as a sequence of base-10000 *digits*.
The first digit is assumed to be multiplied by *weight* * 10000, i.e. there
are weight + 1 digits before the decimal point.  It is possible to have
negative weight.

*dscale*, or display scale, is the nominal precision expressed as number of
base-10 digits after the decimal point.  It is always non-negative.  dscale may
be more than the number of physically present fractional digits, implying
significant trailing zeroes.  It is never less that the number of digits.

For example, the decimal value ``-15000.6250000`` is represented as:

.. code-block:: c

    // ndigits
    0x00 0x04

    // weight
    0x00 0x01

    // sign
    0x40 0x00

    // dscale
    0x00 0x07

    // digits
    0x00 0x01 0x13 0x88 0x18 0x6a 0x00 0x00

.. _ref_protocol_fmt_bool:

std::bool
=========

The :eql:type:`std::bool` values are represented as a single byte value,
``0x01`` for ``true`` and ``0x00`` for ``false``.

.. _ref_protocol_fmt_datetime:

std::datetime
=============

The :eql:type:`std::datetime` values are represented as a 64-bit integer,
most sigificant byte first.  The value is the number of *microseconds*

> between the encoded datetime and January 1st 2000, 00:00 UTC.

Reviewer: Maybe it's worth mentioning that `946684800` is the *proper* number of seconds that should be used to convert between 2000/1970 timestamps.
